rslib/src/scheduler/filtered/custom_study.rs

```diff
@@ -0,0 +1,226 @@
+            }
+            CustomStudyValue::Cram(cram) => {
+                self.create_custom_study_deck(cram_config(current_deck.human_name(), cram)?)
+            }
+        }
+    }
+
+    /// Reuse existing one or create new one if missing.
+    /// Guaranteed to be a filtered deck.
+    fn create_custom_study_deck(&mut self, config: FilteredDeck) -> Result<()> {
+        let mut id = DeckId(0);
+        let human_name = self.tr.custom_study_custom_study_session().to_string();
+
+        if let Some(did) = self.get_deck_id(&human_name)? {
+            if !self
+                .get_deck(did)?
+                .ok_or(AnkiError::NotFound)?
+                .is_filtered()
+            {
+                return Err(CustomStudyError::ExistingDeck.into());
+            }
+            id = did;
+        }
+
+        self.add_or_update_filtered_deck_inner(FilteredDeckForUpdate {
+            id,
+            human_name,
+            config,
+        })
+        .map(|_| ())
+        .map_err(|err| {
+            if err == AnkiError::FilteredDeckError(FilteredDeckError::SearchReturnedNoCards) {
+                CustomStudyError::NoMatchingCards.into()
+            } else {
+                err
+            }
+        })
+    }
+}
+
+fn custom_study_config(
+    deck_name: String,
+    reschedule: bool,
+    nodes: Vec<Node>,
+    order: FilteredSearchOrder,
+    limit: Option<u32>,
+) -> FilteredDeck {
+    let search = concatenate_searches(
+        BoolSeparator::And,
+        nodes,
+        Node::Search(SearchNode::Deck(deck_name)),
+    );
+    FilteredDeck {
+        reschedule,
+        search_terms: vec![FilteredSearchTerm {
+            search,
+            limit: limit.unwrap_or(99_999),
+            order: order as i32,
+        }],
+        delays: vec![],
+        preview_delay: 10,
+    }
+}
+
+fn forgot_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![Node::Search(SearchNode::Rated {
+        days,
+        ease: RatingKind::AnswerButton(1),
+    })];
+    custom_study_config(deck_name, false, nodes, FilteredSearchOrder::Random, None)
+}
+
+fn ahead_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![Node::Search(SearchNode::Property {
+        operator: "<=".to_string(),
+        kind: PropertyKind::Due(days as i32),
+    })];
+    custom_study_config(deck_name, true, nodes, FilteredSearchOrder::Due, None)
+}
+
+fn preview_config(deck_name: String, days: u32) -> FilteredDeck {
+    let nodes = vec![
+        Node::Search(SearchNode::State(StateKind::New)),
+        Node::And,
+        Node::Search(SearchNode::AddedInDays(days)),
+    ];
+    custom_study_config(
+        deck_name,
+        false,
+        nodes,
+        FilteredSearchOrder::OldestReviewedFirst,
+        None,
+    )
+}
+
+macro_rules! state_node {
+    (not $state_kind:expr) => {
+        Node::Not(Box::new(Node::Search(SearchNode::State($state_kind))))
```

Comment:
Neat. This makes me wonder why I boxed the inner argument in the Not case though - presumably it would have worked without the extra layer of indirection.

Reply:
Isn't it necessary because we would have an potentially infinite chain with unknown size otherwise?
I like the macro too, but I guess it's not very scalable. Might be a better idea to impl some traits and use `Node::negated()`.

Reply:
Ah yep, the recursion issue - not sure why that slipped my mind :-)